### PR TITLE
Update JSON handling strategy

### PR DESCRIPTION
**Pull request by znarf**
We had a discussion about that, and it's currently difficult to enforce object or string in inputs.

So we check in formats for now.

### CODE REVIEW

**Comment by benfle** [comment]
The format parsers should take strings as input.

https://github.com/access-watch/access-watch/blob/master/docs/input.md#custom

Which input calls the JSON parser with an object?

**Comment by znarf** [comment]
@benfle HTTP input, especially useful in the batch mode.

**Comment by benfle** [comment]
Yeah we forgot the batch mode when we designed the input/format protocol. I would add support for batch mode in the format parsers then.

The goal of the input/format breakdown was to decouple the transfer of bytes with the parsing of those bytes into a log entity ready to be ingested by the pipeline.

I see 2 directions here:

1. Decide that the orthogonality of input/format does not help and just keep inputs that will implement both the protocol and the format. They eventually can share some common parsing functions in a library.

2. Keep the input/format decomposition but modify the format interface to support batch mode. The format parsers take a string representing one or more log entities and return a List of immutable maps.

**Comment by znarf** [comment]
This idea (2) of enforcing inputs as string, especially for batch, doesn't sounds great.

So, (1) might be the way to go. Not sure of all implications yet.

**Comment by benfle** [comment]
My intent with the input/format decomposition was to separate the protocol from the format. The protocol retrieves a bunch of bytes (a line in a file, a websocket message, a request body...) and delegate the parsing of those bytes to a format parser. Because we will always deal with text, the API actually deal with UTF8 strings and not bytes. But the idea is the same. The input should not assume anything about the bytes.

But apparently the format leaked into various input and I didn't pay enough attention to that:

The websocket input assumes that it gets JSON:

https://github.com/access-watch/access-watch/blob/master/input/websocket.js#L7

The http input also assumes JSON and allows for batches.

So right now the inputs expect parsers like those:

```
file:              string   -> log
syslog:            string   -> log
websocket:         JSON log -> log
http:              JSON log -> log
elasticsearch:     JSON log -> log
```

I propose the following:

1. Get rid of the idea of a required parser (`string -> log`) for each input, it clearly doesn't work for the `elasticsearch` input. For this input, the user can specify a function in the options that will convert a JS object into an immutable log.
2. Decide what level of flexibility we want for the `websocket` input.
3. Decide what level of flexibility we want for the `http` input.

For (2) and (3), I see the following options:

1. Hardcode the format inside the input. We limit the websocket input to connect 2 instance of AW. We limit the HTTP input to receive logs in a specified format.
2. Make some assumptions but give some flexibility. For example, assume JSON but let the user specify the JSON format of individual logs. This is what we have right now.
3. Make the websocket and HTTP input completely configurable and take a parser (string -> log) as input. Then we would implement the JSON parsing, batch... in the format. And the users can reuse those protocols with their custom format.

**Comment by znarf** [comment]
- I'm good if WebSocket behave exactly like Syslog. It's text based, and we should be able to plug the JSON parser. The hints are that we're currently doing the JSON parsing in the input and there is not gonna be batches.
- HTTP should then behave like Elasticsearch I guess. Maybe having a default that just take the input and make it Immutable.

**Comment by benfle** [comment]
I'm ok with websocket because it matches what we already do.

For `http`, I would just parse the JSON sent and completely bypass the extra flexibility of a user-defined log parser. We can always introduce later if needed.

**Comment by benfle** [comment]
And as I said in the other thread, we should get rid of the `format/json` module.

File: input/http.js [hunk 1]
```diff
@@ -11,8 +13,11 @@ function create ({name = 'HTTP endpoint', endpoint, parse}) {
         let messages = Array.isArray(req.body) ? req.body : [req.body]
         messages.forEach(message => {
           try {
-            const log = parse(req.body)
-            pipeline.success(log)
+            if (parse) {
```
Comment (benfle) [inline input/http.js]: Why don't you just set a default value for `parse` to `fromJS`?

Comment (znarf) [reply input/http.js]: that could work there, but for the text based inputs, that is ```fromJS(JSON.parse(msg))``` so less simple.

**Comment by znarf** [comment]
What about now with the latest commit? https://github.com/access-watch/access-watch/pull/39/commits/da744a7032ad45a2e5f827ad7bc2b65d886a21e4

**Comment by benfle** [comment]
👍

**Pull request merged**